### PR TITLE
Log payload only [logentries]

### PR DESCRIPTION
**Pull request by sascha-andres**
If you want to use the advanced query in logentries
logging in json is very useful. If you do not need
the advanced information with
--log-opt logentries-line-only=true just logs the payload

Created issue for enhancement: #31726

### CODE REVIEW

**Comment by cpuguy83** [comment]
It may be helpful to have this available to all drivers.
Particularly if people are already using structured logging.

**Comment by thaJeztah** [comment]
That sounds similar to https://github.com/docker/docker/issues/24898, https://github.com/docker/docker/issues/17020 ?

**Comment by sascha-andres** [comment]
#24652 implements this for splunk when a logging option with value raw is set. It seems that similar requirements pop up for different logging drivers. Basically, wrapping into a serialized go map makes it harder for our devs to analyse logs generated by apache or php within the containers. Therefore I came up with this and have not found something like this for logentries

**Comment by kevin-bockman** [comment]
I am using Logentries as well and am running into this same problem. This PR seemed to have stalled. Can this be merged?

**Comment by LK4D4** [comment]
For move this forward I propose to just use generic `line-only` name(without logentries part), but implement only for logentries.
@sascha-andres @vdemeester @cpuguy83

**Comment by sascha-andres** [comment]
@LK4D4 Adjusted from logentries-line-only to line-only

File: daemon/logger/logentries/logentries.go [hunk 1]
```diff
@@ -38,32 +41,46 @@ func init() {
...
 		WithField("token", info.Config[token]).
+		WithField("line-only", info.Config[lineonly]).
 		Debug("logging driver logentries configured")
 
 	log, err := le_go.Connect(info.Config[token])
 	if err != nil {
 		return nil, err
 	}
+	var lineOnly bool
+	if "" == info.Config[lineonly] {
```
Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]:
```
if _, exists := info.Config[lineOnly]; exists {
  // do strconv
}
```

There is no need to explicitly set false.
I think we should error out if nothing is set, which will happen with `ParseBool`

File: daemon/logger/logentries/logentries.go [hunk 2]
```diff
@@ -38,32 +41,46 @@ func init() {
...
+		lineOnly = false
+	} else {
+		if lineOnly, err = strconv.ParseBool(info.Config[lineonly]); err != nil {
+			return nil, err
```
Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: Please use `errors.Wrap` to provide context to the error.

**Comment by cpuguy83** [comment]
Design OK.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Review by sascha-andres — COMMENTED** [review]
@cpuguy83 Changed implementation

File: daemon/logger/logentries/logentries.go [hunk 3]
```diff
@@ -38,32 +42,42 @@ func init() {
...
+		data := map[string]string{
+			"container_id":   f.containerID,
+			"container_name": f.containerName,
+			"source":         msg.Source,
+			"log":            string(msg.Line),
+		}
+		for k, v := range f.extra {
+			data[k] = v
+		}
+		logger.PutMessage(msg)
```
Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: We can't `PutMessage` until all the values are copied.
I think in this case we just need to copy `msg.TimeStamp`

**Comment by LK4D4** [comment]
@sascha-andres would you mind to fix comments? Thanks!

**Comment by sascha-andres** [comment]
@LK4D4 I will take a look and report back.

File: daemon/logger/logentries/logentries.go (hunk 3 above)
Comment (sascha-andres) [inline daemon/logger/logentries/logentries.go]: Just to be sure about this. The original source code looks like this:

```
data := map[string]string{
    "container_id":   f.containerID,
    "container_name": f.containerName,
    "source":         msg.Source,
    "log":            string(msg.Line),
}
for k, v := range f.extra {
    data[k] = v
}
ts := msg.Timestamp // ts used for Printf later, I used msg.Timestamp directly
logger.PutMessage(msg)
```
You are proposing adding the timestamp to data in addition to the one included in msg?

Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: That seems like it would work since we aren't actually modifying the timestamp itself at any point.

Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: Sorry, let me clarify, the current code is doing a `PutMessage`, as such you can't use `msg.Timestamp` reliably below in the `Println`

You will have to `ts := msg.Timestamp // ts used for Printf later, I used msg.Timestamp directly` and use `ts`.

File: daemon/logger/logentries/logentries.go [hunk 4]
```diff
@@ -38,32 +42,42 @@ func init() {
...
+		f.writer.Println(f.tag, msg.Timestamp, data)
+	} else {
+		logger.PutMessage(msg)
```
Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: Likewise here you can't `PutMessage` and then use `msg.Line`.

File: daemon/logger/logentries/logentries.go (hunk 3 above)
Comment (sascha-andres) [inline daemon/logger/logentries/logentries.go]: Thanks for your explanation. Changed the code to reflect my new understanding

Comment (sascha-andres) [inline daemon/logger/logentries/logentries.go]: Thanks for your explanation. Changed the code to reflect my new understanding

**Comment by thaJeztah** [comment]
@sascha-andres looks like this needs a rebase due to the `github.com/Sirupsen/logrus` -> `github.com/sirupsen/logrus` rename; can you also squash your commits?

**Comment by sascha-andres** [comment]
@thaJeztah Will squash and rebase

File: daemon/logger/logentries/logentries.go [hunk 5]
```diff
@@ -38,32 +42,44 @@ func init() {
...
 		Debug("logging driver logentries configured")
 
 	log, err := le_go.Connect(info.Config[token])
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "error connecting to logentries")
+	}
+	var lineOnly bool
+	if lineOnly, err = strconv.ParseBool(info.Config[lineonly]); err != nil {
+		return nil, errors.Wrap(err, "error parsing lineonly option")
```
Comment (cpuguy83) [inline daemon/logger/logentries/logentries.go]: nit `line-only`

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Pull request merged**